iree/base/wait_handle_test.cc

```diff
@@ -14,538 +14,803 @@
-  WaitHandle wh0 = fence0.OnSet();
-  WaitHandle wh1 = fence1.OnSet();
-  IREE_ASSERT_OK(WaitHandle::WaitAll({&wh0, &wh1}));
-
-  t0.join();
-  t1.join();
+TEST(Event, WaitOneInitialTrue) {
+  iree_event_t event;
+  IREE_ASSERT_OK(iree_event_initialize(/*initial_state=*/true, &event));
+  IREE_EXPECT_OK(iree_wait_one(&event, IREE_TIME_INFINITE_PAST));
+  iree_event_deinitialize(&event);
 }
 
-// Tests using WaitAll with multiple wait handles from the same fence.
-TEST(WaitHandleTest, WaitAllSameSource) {
-  ManualResetEvent fence;
-  WaitHandle wh0 = fence.OnSet();
-  WaitHandle wh1 = fence.OnSet();
-  ASSERT_TRUE(
-      IsDeadlineExceeded(WaitHandle::WaitAll({&wh0, &wh1}, InfinitePast())));
-  IREE_ASSERT_OK(fence.Set());
-  IREE_ASSERT_OK(WaitHandle::WaitAll({&wh0, &wh1}));
-}
+TEST(Event, SetWait) {
+  iree_event_t event;
+  IREE_ASSERT_OK(iree_event_initialize(/*initial_state=*/false, &event));
 
-// Tests using WaitAll with literally the same wait handles.
-TEST(WaitHandleTest, WaitAllSameHandle) {
-  ManualResetEvent fence;
-  WaitHandle wh = fence.OnSet();
-  ASSERT_TRUE(
-      IsDeadlineExceeded(WaitHandle::WaitAll({&wh, &wh}, InfinitePast())));
-  IREE_ASSERT_OK(fence.Set());
-  IREE_ASSERT_OK(WaitHandle::WaitAll({&wh, &wh}));
-}
+  // Initially unset.
+  IREE_EXPECT_STATUS_IS(IREE_STATUS_DEADLINE_EXCEEDED,
+                        iree_wait_one(&event, IREE_TIME_INFINITE_PAST));
 
-// Tests WaitAll when a wait handle fails.
-TEST(WaitHandleTest, WaitAllFailure) {
-  WaitHandle good_wh;
-  // Create a purposefully bad handle to induce an error.
-  WaitHandle bad_wh = WaitHandle::AlwaysFailing();
-  // Should fail with some posixy error.
-  ASSERT_FALSE(WaitHandle::WaitAll({&good_wh, &bad_wh}).ok());
-}
+  // Set and wait.
+  iree_event_set(&event);
+  IREE_EXPECT_OK(iree_wait_one(&event, IREE_TIME_INFINITE_PAST));
+
+  // Set should be sticky until reset manually.
+  IREE_EXPECT_OK(iree_wait_one(&event, IREE_TIME_INFINITE_PAST));
 
-// Tests using WaitAny with no valid handles. This should no-op.
-TEST(WaitHandleTest, WaitAnyNop) {
-  ASSERT_TRUE(IsInvalidArgument(WaitHandle::WaitAny({}).status()));
-  IREE_ASSERT_OK_AND_ASSIGN(int index, WaitHandle::WaitAny({nullptr}));
-  ASSERT_EQ(0, index);
-  IREE_ASSERT_OK_AND_ASSIGN(index, WaitHandle::WaitAny({nullptr, nullptr}));
-  ASSERT_EQ(0, index);
+  // Resetting should unsignal the event.
+  iree_event_reset(&event);
+  IREE_EXPECT_STATUS_IS(IREE_STATUS_DEADLINE_EXCEEDED,
+                        iree_wait_one(&event, IREE_TIME_INFINITE_PAST));
+
+  iree_event_deinitialize(&event);
 }
 
-// Tests polling with WaitAny with multiple wait handles.
-TEST(WaitHandleTest, WaitAnyPoll) {
-  ManualResetEvent fence0;
-  WaitHandle wh0 = fence0.OnSet();
-  ManualResetEvent fence1;
-  WaitHandle wh1 = fence1.OnSet();
+// Tests that we can use set/reset and that certain behavior (such as sets
+// without intervening resets) is allowed. Note that this does not wait and is
```

Comment:
A bit weary about ASAN/MSAN/TSAN and other similar checks in this file. I'd say go with what our open source CI is actually checking, but someone might need to triage some warnings/errors after merge.

Reply:
A lot of the patterns in this file originated from google3 where asan/msan/etc was being used. Still possible for there to be issues as the implementation is really different, but that's even more reason that someone should set up a bot :P